extension/sigv4authextension/design.md

```diff
@@ -0,0 +1,559 @@
+    } else {
+        ua = si.awsSDKInfo
+    }
+    req2.Header.Set("User-Agent", ua)
+
+	// Hash the request
+	h := sha256.New()
+	_, _ = io.Copy(h, body)
+	payloadHash := hex.EncodeToString(h.Sum(nil))
+
+	// Use user provided service/region if specified, use inferred service/region if not, then sign the request
+	service, region := si.inferServiceAndRegion(req)
+	creds, err := (*si.credsProvider).Retrieve(req.Context())
+	if err != nil {
+		return nil, errBadCreds
+	}
+	err = si.signer.SignHTTP(req.Context(), creds, req2, payloadHash, service, region, time.Now())
+	if err != nil {
+		return nil, fmt.Errorf("error signing the request: %w", err)
+	}
+
+	// Send the request
+	return si.transport.RoundTrip(req2)
+} 
+```
+
+
+**Performance Considerations of `RoundTrip()`**
+
+
+We take a closer look at the performance of `RoundTrip()`, since it will be heavily used i.e. called by every HTTP transaction, and can also be used for exporting to any AWS service. Our goal here is to outline both the memory and runtime performance of this method, as well as explain design choices.
+
+
+```go
+    reqBody, err := req.GetBody()
+    .
+    .
+    .
+    content, err := ioutil.ReadAll(reqBody)
+    reqBody.Close()
+    .
+    .
+    .
+    body := bytes.NewReader(content)
+```
+
+
+First, we obtain the body. This should not have an impact on runtime performance or memory performance as we do not expect to handle http requests that are exponentially large in size*. `body` is necessary as it is passed into the `Sign()` method later.
+
+
+**Note that OTEL does not limit the size of HTTP requests, except if the exporters themselves set a limit. No information could be found on the size of AWS API requests either. If the request body is exponentially large it could affect performance but it is a reasonably safe assumption that that will not be the case.*
+
+
+```go
+    req2 := cloneRequest(req)
+    .
+    .
+    .
+    ua := req2.Header.Get("User-Agent")
+    ua = ua + " " + si.awsSDKInfo // or ua = si.awsSDKInfo
+    req2.Header.Set("User-Agent", ua)
+```
+
+
+Next, we clone the request and also add runtime information to the User-Agent header. `cloneRequest()` is a helper function that makes a shallow copy of the request and a deep copy of the header. While this may slow the performance of `RoundTrip()`, it is necessary to ensure thread safety*. Overall, this should not have a large impact on either runtime or memory performance, as again we do not expect to handle http requests that are exponentially large in size. Adding runtime information to the header also will not impact memory or runtime performance either. `req2` will be passed into `Sign()`.
+
+
+**Design Decision: The necessity of `cloneRequest()` is unknown; for now we will leave it in, as this implementation of `RoundTrip()` is taken directly from the AWS PRW Exporter. However, if we want to avoid reading the body, that can only happen if we avoid using `cloneRequest()`.*
```

Comment:
Isn't a request object immutable? Perhaps that's why it's cloning?

Reply:
This makes sense, @Aneurysm9 what do you think? Will wait on rewriting this for now, but in the meantime will reword the last sentence to read "However, if we want to avoid using `cloneRequest()`, that can only be done by avoiding reading the body".

Reply:
SGTM.  We could perhaps be a bit more forceful with that statement since it is impossible to avoid reading the body, but I'm not sure it makes a huge difference either way.